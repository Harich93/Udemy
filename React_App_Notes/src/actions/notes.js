
import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const stratNewNote = () => {
    return async ( dispatch, getState ) => {

        const { uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await db.collection( `${ uid }/journal/notes` ).add( newNote )
        
        dispatch( activeNote( doc.id, newNote) );
        dispatch( addNewNote( doc.id, newNote ) );
    }

}
 
export const activeNote = ( id, note ) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNewNote =( id, note ) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = ( uid ) => {
    return async( dispatch ) => {
        const notes = await loadNotes( uid );
        dispatch( setNotes( notes ) );
    }
}

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
});

export const savedNote = ( note ) => {

    return async( dispatch, getState ) => {

        const { uid } = getState().auth;
        const noteToFirestore = { ...note };
        delete noteToFirestore.id;
        // !note.url && delete note.url
        

        const arch = await db.doc( `${ uid }/journal/notes/${ note.id }` )
            .update( noteToFirestore )

        console.log( arch )

        dispatch( refreshNote( note.id, note ) );
        Swal.fire( 'Saved', note.title, 'success');
    } 
}

export const refreshNote = ( id, note ) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = ( file ) => {
    return async( dispatch, getState ) => {
        const { active: activeNote } = getState().note;
        Swal.fire({
            title: 'Uploading...',
            text: 'Please await...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            },   
        });
        
        const fileUrl = await fileUpload( file );
        activeNote.url = fileUrl;
        
        dispatch( savedNote( activeNote ) );

        Swal.close();
    }
}

export const startDeleting = ( id ) => {
    console.log(id)
    return async( dispatch, getState ) => {
        
        const { uid } = getState().auth;
        
        await db.doc( `${ uid }/journal/notes/${ id }` ).delete()
            .then( () => Swal.fire( 'Delete', id, 'success') )
            .catch( console.log )

        dispatch( deleteNote(id) ); 
    }

}

export const deleteNote = ( id ) => ({
    type: types.notesDeleted,
    payload: id
})

export const logoutNote = () => ({
    type: types.notesLogoutCleaning
})