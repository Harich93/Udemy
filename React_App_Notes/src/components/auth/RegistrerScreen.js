import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { setError, removeError} from '../../actions/ui'
import { startRegisterWhithEmailPassword } from '../../actions/auth';


export const RegistrerScreen = () => {

    const dispatch = useDispatch();

    const { msgError } = useSelector(  state => state.ui )

    const [ values, handleInputChange ] = useForm({ 
        name: '',
        email: '',
        password: '',
        password2: ''
    });

    const { name, email, password, password2 } = values;

    const handleRegister = ( e ) => { 
        e.preventDefault();

        if ( isFormValid() ) {
            dispatch( startRegisterWhithEmailPassword( email, password, name ) );
        }
    }

    const isFormValid = () => {

        if( name.trim().length <= 1 ) {
            dispatch( setError('Name is required') )
            return false
        } else if ( !validator.isEmail( email ) ) {
            dispatch( setError('Email is not valid') );
            return false;
        } else if ( password !== password2 || password.length < 5 ) {
            dispatch( setError('Password should be at least 6 characters and match each other') );
            return false;
        }

        dispatch( removeError() );
        return true;
        
    }

    return (

        <>

            <h3 className='auth__title'>Register</h3>

            {
                msgError &&
                (
                    <div className='auth__alert-error'>
                        { msgError }
                    </div>
                )

            }

            <form 
                onSubmit={ handleRegister }
                className='animate__animated animate__fadeIn animate__fast'    
            >

                <input
                    type='text'
                    placeholder='Name'
                    name='name'
                    value={ name }
                    autoComplete='off'
                    className='auth__input'
                    onChange={ handleInputChange }
                    
                />

                <input
                    type='text'
                    placeholder='Email'
                    name='email'
                    value={ email }
                    autoComplete='off'
                    className='auth__input'
                    onChange={ handleInputChange }
                    
                />

                <input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    value={ password }
                    className='auth__input'
                    onChange={ handleInputChange }
                />

                <input 
                    type='password'
                    placeholder='Confirm password'
                    name='password2'
                    value={ password2 }
                    className='auth__input'
                    onChange={ handleInputChange }
                />

                <button
                    type='submit'
                    className='btn btn-primary btn-block mb-5 mt-1'
                >Register</button>

                <Link className='link' to='/auth/login' >
                    Already registered?
                </Link>
            </form>
        </>
    )
}
