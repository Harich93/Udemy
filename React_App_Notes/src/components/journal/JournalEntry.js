import moment from 'moment'
import React from 'react'
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';


export const JournalEntry = ( note ) => {

    const {id, body, date, title, url } = note

    const dispatch = useDispatch()

    const noteDate = moment(date);
    
    const handleEditNote = () => {
        dispatch( activeNote( id, note ) );
    }

    return (
        <div 
            className='journal__entry pointer animate__animated animate__fadeIn'
            onClick={ handleEditNote }
        >

            {
                url &&
                    <div
                        className='journal__entry-picture'
                        style={ { 
                            backgroundImage: `URL(${ url })`, 
                            backgroundSize: 'cover'
                        } }
                    ></div>
            }


                <div className='journal__entry-body'>
                    <p className='journal__entry-title'>  
                        { title }
                    </p>
                    <p className='journal__entry-content'>  
                        { body }
                    </p>
                </div>

                <div className='journal__entry-date-box'>
                    <span>{ noteDate.format('dddd') }</span>
                    <h4>{ noteDate.format('Do') }</h4>
                </div>

        </div>
    )
}
