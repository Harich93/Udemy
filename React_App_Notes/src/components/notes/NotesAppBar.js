import moment from 'moment'
import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { savedNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {
    
    const date = moment().format( 'LL' );

    const note = useSelector(state => state.note.active );
    
    const dispatch = useDispatch();
    
    const handleSave = () => {

        dispatch( savedNote( note ) );
    }

    const handleUpdatePicture = () => {
        document.querySelector('#fileSelector').click();
    }
    
    const handleFileChanged = ( e ) => {
        const file = e.target.files[0];

        file && dispatch( startUploading( file ) );

    }

    return (
        <div className='notes__appbar'>
            <span>{ date }</span>

            <input
                id='fileSelector'
                name='file'
                type='file'
                style={ { display: 'none' } }
                onChange={ handleFileChanged }
            />

            <div>
                <button 
                    className='btn'
                    onClick={ handleUpdatePicture }    
                > 
                    Picture
                </button>

                <button 
                    className='btn'
                    onClick={ handleSave }
                > 
                    Save
                </button>
            </div>
        
        </div>
    )
}
