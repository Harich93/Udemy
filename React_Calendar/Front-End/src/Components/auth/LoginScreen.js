import React from 'react';
import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';
import { startLogin, startRegister } from '../../Actions/auth';
import { useForm } from '../../Hooks/useForm'
import { useViewPass } from '../../Hooks/useViewPass';
import '../../Style/login.css';

export const LoginScreen = () => {

    const dispatch = useDispatch();

    const [ isView, handleChangeViewPass ] = useViewPass();
    const [ isViewLogin, handleChangeViewPassLogin ] = useViewPass();


    const [ formLoginValues, handleLoginInputChange ] = useForm({
        lEmail: '',
        lPassword: ''
    });
    
    const { lEmail, lPassword } = formLoginValues;

    const [ formRegisterValues, handleRegisterInputChange ] = useForm({
        rName: '',
        rEmail: '',
        rPassword: '',
        rPassword2:''
    });

    const { rName, rEmail, rPassword, rPassword2 } = formRegisterValues;


    const handleLogin = (e) => {
        e.preventDefault();
        dispatch( startLogin( lEmail, lPassword ) );
    }

    const handleRegister = (e) => {
        e.preventDefault();
        console.log(rPassword, rPassword2)
        if( rPassword !== rPassword2 )
            return Swal.fire('Error','The passwords must be equal','error')
        
        dispatch( startRegister( rEmail, rPassword, rName ) );
    }

    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Ingreso</h3>
                    <form onSubmit={ handleLogin }>
                        <div className="form-group">
                            <input 
                                type="text"
                                className="form-control"
                                placeholder="Correo"
                                name='lEmail'
                                onChange={ handleLoginInputChange }
                                value={ lEmail }
                            />
                        </div>
                        <div className="form-group">
                            <div className='input-group'>
                                <input
                                    type={ isViewLogin ? "text" : "password"}
                                    className="form-control pass"
                                    placeholder="Contraseña" 
                                    name='lPassword'
                                    value={ lPassword }
                                    onChange={ handleLoginInputChange }
                                />
                                <span className='input-group-addon' onClick={ handleChangeViewPassLogin }>
                                    {
                                        isViewLogin
                                            ? <i className="fas fa-eye-slash"></i>
                                            : <i className="fas fa-eye"></i>
                                    }
                                </span>

                            </div>
                        </div>
                        <div className="form-group">
                            <input 
                                type="submit"
                                className="btnSubmit"
                                value="Login" 
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Registro</h3>
                    <form onSubmit={ handleRegister }>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Nombre"
                                name='rName'
                                value={ rName }
                                onChange={ handleRegisterInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Correo"
                                name='rEmail'
                                value={ rEmail }
                                onChange={ handleRegisterInputChange }
                            />
                        </div>
                        <div className="form-group">
                            <div className='input-group'>
                                <input
                                    type={ isView ? "text" : "password"}
                                    className="form-control pass"
                                    placeholder="Contraseña" 
                                    name='rPassword'
                                    value={ rPassword }
                                    onChange={ handleRegisterInputChange }
                                />
                                <span className='input-group-addon' onClick={ handleChangeViewPass }>
                                    {
                                        isView
                                            ? <i className="fas fa-eye-slash"></i>
                                            : <i className="fas fa-eye"></i>
                                    }
                                </span>

                            </div>
                        </div>

                        <div className="form-group">
                            <input
                                type={ isView ? "text" : "password"}
                                className="form-control"
                                placeholder="Repita la contraseña"
                                name='rPassword2'
                                value={ rPassword2 }
                                onChange={ handleRegisterInputChange } 
                                />
                        </div>

                        <div className="form-group">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Crear cuenta" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}
