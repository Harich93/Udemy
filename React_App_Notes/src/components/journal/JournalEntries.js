import React from 'react'
import { useSelector } from 'react-redux'
import { JournalEntry } from './JournalEntry'

export const JournalEntries = () => {
    
    const { notes }  = useSelector(state => state.note)

    return (
        <div className='journal__entries  animate__animated animate__fadeInLeft animate__slow'>
            {
                notes.map( note => (
                    <JournalEntry
                        key={ note.id } 
                        { ...note }
                    />
                ))
            }
        </div>
    )
}
