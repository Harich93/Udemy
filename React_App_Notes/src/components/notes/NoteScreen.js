import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const note = useSelector(state => state.note.active );
    const [ values, handleInputChange, reset ] = useForm(note);
    const { title, body, url } = values;

    const activeId = useRef( note.id );
    const activeUrl = useRef( note.url );

    const handleDelete = () => {
        dispatch( startDeleting( note.id ) );
    }
    
    useEffect(() => {

        if ( note.id !== activeId.current ) {
            reset( note );
            activeId.current = note.id;
        }
        if ( note.url !== activeUrl.current ) {
            reset( note );
            activeUrl.current = note.url;
        }

    }, [note, reset, note.url ]);

    useEffect(() => {
        
        dispatch( activeNote( values.id, { ...values } ) );

    }, [values, dispatch,] )
    
    return (
        <div className='notes___main-content'>
            <NotesAppBar />

            <div className='notes__content'>

                    <input 
                        type='text'
                        name='title'
                        placeholder='Some awasome title'
                        className='notes__title-input'
                        value={ title }
                        autoComplete='off'
                        onChange={ handleInputChange }
                    />

                    <textarea
                        placeholder='What happened today'
                        name='body'
                        className='notes__textarea'
                        value={ body }
                        onChange={ handleInputChange }
                    ></textarea>

                    {
                        url &&
                            <div className='notes__image'>
                                <img
                                    src={ url }
                                    alt='Imagen nota'
                                />
                            </div>
                    }

             
            </div>
            <button
                className='btn btn-delete'
                onClick={ handleDelete }
            >Delete</button>
        </div>
    )
}
