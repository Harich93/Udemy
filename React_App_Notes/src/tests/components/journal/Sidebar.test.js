import '../../../setupTest'

import React from 'react';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store'
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import { Sidebar } from '../../../components/journal/Sidebar';
import { logoutNote } from '../../../actions/notes';
import { startLogout } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
    startLogout: jest.fn(),
}))
jest.mock('../../../actions/notes', () => ({
    logoutNote: jest.fn(),
}))


const middlewares = [thunk];
const mockStore = configureStore( middlewares );

const initialState = {
    auth:{
        name: 'name'
    },
    note: {
        notes: [],
        active: false
    }
}

let store = mockStore( initialState );
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store } >
        <MemoryRouter>
            <Sidebar />
        </MemoryRouter>
    </Provider>
)

describe('Pruebas en <Sidebar />', () => {
    
    test('Debe mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('Debe de llamar a startLogout y logoutNote', () => {
        
        wrapper.find('button').prop('onClick')();

        expect( startLogout ).toHaveBeenCalled();
        expect( logoutNote ).toHaveBeenCalled();


    });
});