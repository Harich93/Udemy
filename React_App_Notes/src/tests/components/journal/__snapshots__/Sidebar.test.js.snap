// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Sidebar /> Debe mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar mt-3"
  >
    <h3>
      <i
        className="far fa-user-circle fs-2"
      />
      <span
        className="fs-2"
      >
         
        name
         
      </span>
    </h3>
    <button
      className="btn btn-logout"
      onClick={[Function]}
    >
      Logout 
      <i
        className="fas fa-sign-out-alt fs-1"
      />
    </button>
  </div>
  <div
    className="journal__new-entry"
  >
    <i
      className="far fa-calendar-plus fa-5x"
      onClick={[Function]}
    />
    <p
      className="mt-5"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries  animate__animated animate__fadeInLeft animate__slow"
  />
</aside>
`;
